src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    /*
+     * Implementation note
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
+     * changes)
+     * Case 2: we have already synced (once or more) during the current
+     * term, but need to sync again (e.g. re-starting the archiver to apply
+     * topic configs)
+     *
+     * They are handled separately to the reader's delight. In theory, it is
+     * possible to fallthrough from Case 1 to Case 2. This function should be
+     * called in a loop until it succeeds.
+     */
+    if (!_raft->is_leader()) {
+        co_return false;
+    }
+
+    const auto current_term = _raft->term();
+    const auto last_applied_offset = _manifest->get_insync_offset();
+
+    if (current_term == _insync_term && _last_replicated_offset) {
+        if (last_applied_offset < _last_replicated_offset) {
```

Comment:
Given how the value is updated this can only be the case if we're calling this method concurrently. But this means that we have two active `ntp_archiver` instances. One is replicating something and about to step down and another one is starting uploads. We assume that there is always only one `ntp_archiver` and it's either reading data from the STM or replicating something.

Reply:
You're right. The place where `last_replicated_offset` is applied is not correct. Thanks for pointing out. Will fix.

Reply:
Fixed